### ISSUE TITLE
SYS without arguments

### ISSUE THREAD

**Issue by mist64** [commanderx16/x16-rom#107]
We should support a version of the BASIC SYS statement without arguments that will execute machine language code that follows the BASIC program in RAM.

Background: Currently, the X16's BASIC RAM starts at $0800 to avoid linking problems with BASIC programs loaded from the C64. I'd like to change this to $0400 though, because we have no other use for the RAM at $0400-$07FF, and it'll give BASIC another KB to work with. A change like this will break "10SYS2061"-style BASIC headers. These headers would also be incompatible between different Commodore-like machines.

So I'm proposing allowing SYS without arguments. In this case, it will jump to the address stored in "vartab", which points to the end of the BASIC program. This allows BASIC headers like this:

10 SYS:REM (C)2019 ACME INC.

**Comment by Kroc** [comment]
I commented somewhere in a line-comment, that the $0400-$0800 space would be useful for once-off initialisation code, to be reused by the user-program as variable space. Like, for example when we boot-strap C64 programs at $0334.

If BASIC starts at $0400, you have less flexibility for bootstrap / initialisation and the main program code. A small bootstrap that fits below $0400 may well do some configuration and then load a program from disk into $0800+

**Comment by mist64** [comment]
Good point. I've been missing the C64's $C000-$CFFF space on the X16. $0400-$07FF could be this. Scratch RAM that can be used for machine code helpers, temporary storage etc.

**Comment by BruceMcF** [comment]
I'm trying to make my Forth leave a 2K space for Basic in place, for system file handling & other things, and so have been planning on moving my return stack, input buffer, <S S> number conversion space, PAD and single tasking space for user variables into $0400-$07FF.

And I would also definitely use it as Golden RAM for ML assisted Basic Programming.

Moving Basic doesn't strictly speaking require this change ... if the location is stable after it's been moved. But any option to have Basic start at a specified location would. And it simplifies the Basic stub.

So while I'm not a big fan of moving Basic to $0401, I also support this extension of the SYS keyword.

**Comment by mist64** [comment]
I've documented $0400-$07FF as "available for machine code programs or custom data storage" now.

The plans for `SYS` are still on. :)

**Comment by Kroc** [comment]
This may require that BASIC programs written on the system append some kind of error handler (a null value?) at the end of the BASIC program, so that calls to `SYS` will not crash the machine.

**Comment by BruceMcF** [comment]
> This may require that BASIC programs written on the system append some kind of error handler (a null value?) at the end of the BASIC program, so that calls to `SYS` will not crash the machine.

It only requires that if there is no executable code following the Basic, it doesn't call SYS without an operand. A Basic program would still be free to, eg, call Kernel functions, or ML routines it had stuffed or loaded into $0400-$07FF.

**Comment by oziphantom** [comment]
worth noting that different Commodores have different starts positions. 
VIC-20 $1000
64 is $0801
+4 is $1000/$4000
128 is $1C01

**Comment by BruceMcF** [comment]
> 
> 
> worth noting that different Commodores have different start positions.
> VIC-20 $1000
> VIC-20 Expanded $1200
> 64 is $0801
> +4 is $1000/$4000
> 128 is $1C01

Quite ... while the technique was applicable, the specific sequence to put a basic line to sys to a following binary executable was customized to the specific platform.
